### DISCUSSION TITLE
Improve machine learning resilience in low VRAM scenarios (CUDA)

### DISCUSSION THREAD

**Discussion by flobernd** [immich-app/immich#11981] [Feature Request]
### The bug

Hi there,

I'm currently evaluating Immich and really liking it so far. My Immich instance is running via Docker on a Debian 12 VM that is hosted on ESXi. The VM has a vGPU profile with 4 GiB VRAM assigned.

During a stress-test of the hardware accelleration (both, transcoding and machine learning), I noticed that the machine learning Python module does not seem to be very resilient against low VRAM scenarios.

After uploading some initial photos and videos, I run a "stress-test" by starting re-running all relevant processing tasks simultaneously (face detection, smart search, transcoding). At the same time, I ran a smart search query from the main dashboard.

The following observations were made:

- Especially the smart search query allocates a lot of VRAM
- When the smart search query request fails due to low memory:
  1. A corresponding exception (failed to allocate memory) is logged in the container
  2. The smart search container seems to **leak** the VRAM (it's only released when the container gets restarted)
  3. The smart search container does not seem to continue processing the remaining assets (face recognition, smart search)
  4. Not even re-running the smart search job for all assets helps in this case (only after a container restart)

It would be nice, if the machine learning code could gracefully handle the "failed to allocate memory" exception by releasing all accquired VRAM. A retry mechanic could be implemented to ensure proper processing of assets. If the smart search query fails due to low memory, a specialized message should be displayed in the UI that asks the user to retry later.

### The OS that Immich Server is running on

Debian 12 on ESXi

### Version of Immich Server

v1.112.1

### Version of Immich Mobile App

-

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
-
```

### Your .env content

```Shell
-
```

### Reproduction steps

```bash
See bug description.
```

### Relevant log output

```
[E:onnxruntime:, sequential_executor.cc:516 ExecuteKernel] Non-zero status code returned while running FusedConv node. Name:'Conv_6' Status Message: /onnxruntime_src/onnxruntime/core/framework/bfc_arena.cc:376 void* onnxruntime::BFCArena::AllocateRawInternal(size_t, bool, onnxruntime::Stream*, bool, onnxruntime::WaitNotificationFn) Failed to allocate memory for requested buffer of size 22937600
```

### Additional information

NVIDIA Gridd Driver Version: 550.90.07
CUDA Version: 12.4
vGPU Profile: GRID T4-4Q

**Comment by mertalev** [comment]
VRAM gets freed if the model fails to load, but in this case it seems to fail during inference after it's loaded. This is kind of an edge case where 4GB is *just* short of being enough. I don't think this is a memory leak, though. If another model gets unloaded, the failing model should run inference just fine; it just doesn't have enough resources for inference at that point.

Handling this by removing the model that ran out of memory is a bit sketchy. If you do a text search before uploading an asset (which will load the other models), then one queue will repeatedly load a model from disk, try running inference, fail and repeat. If this edge case occurs just with jobs without the text model, the models will similarly keep getting unloaded and reloaded.

There's room for improvement here, but I think this is more of an enhancement request than a bug.

**Reply by flobernd** [reply]
Hi @mertalev, 

thanks for the explanation! I incorrectly assumed that a complete model failed to load. If it's just the inference itself that does not have enough resources, it of course does not make sense to unload other models.

It might still make sense to unload the text search model if one of the ingestion/processing jobs fails due to low memory. Ingestion (except for the initial run maybe) is a rather short living task and a very important one. It should have precedence over a text search.

I think this should still be classified as a bug as the machine learning container does not seem to recover after the memory allocation failure. Btw. I started the jobs first and ran a search query afterwards. This caused the smart search job to stop in the middle of processing. It never continues in this case and even re-running the job on all assets has no effect in this case.

**Reply by mertalev** [reply]
> I think this should still be classified as a bug as the machine learning container does not seem to recover after the memory allocation failure

A model is unloaded after 5 minutes of not getting a request. If the service doesn't get any requests for 5 minutes after getting an inference request, it restarts with a new process to make sure all RAM and VRAM is freed.

**Reply by flobernd** [reply]
@mertalev Ok, this could have been the problem, as I've periodically re-tried the requests after a while to see if the container handles the requests again.

I can very much live with a few failing search queries 🙂 

Would you say that it's safe to assume that the ingestion jobs will pick up every single assets after a while, or is that not guaranteed if the processing fails with a low memory condition?

**Reply by mertalev** [reply]
> Would you say that it's safe to assume that the ingestion jobs will pick up every single assets after a while, or is that not guaranteed if the processing fails with a low memory condition?

Jobs won't retry automatically unfortunately. This is something we intend to do in the future though.

**Reply by flobernd** [reply]
Ok, thanks again @mertalev. In this case I'm assigning a 8 GiB vGPU profile for now, until automatic retry got implemented.

**Reply by mertalev** [reply]
8 GiB is quite a bit more than necessary I think. If you're allocating that much, you may as well change the search model for a [nicer one](https://huggingface.co/collections/immich-app/clip-654eaefb077425890874cd07). `ViT-L-16-SigLIP-384__webli` is one of the better models and should fit comfortably.

**Reply by flobernd** [reply]
Thanks for the hint 🙂 

I switched to immich-app/XLM-Roberta-Large-Vit-B-16Plus for multi language support.

Unfortunately the vGPU slices are all powers of 2. I can assign either 1, 2, 4, 8 or 16, but nothing inbetween. For that reason it would have been cool if 4 GiB would have worked out, but 8 is fine for now as well. Have to look into that again when I really need the resources for some other VM.
